### PR TITLE
Samza runner support for non unique stateId across multiple ParDos

### PR DESCRIPTION
**Pull request by seungjin-an**
Samza runner support for non-unique stateIds across multiple ParDos
(1) Do not touch existing RocksDB Store Id with stateId used only by one ParDo
(2) Add a prescan logic in SamzaRunner to traverse Beam pipeline topologically and identify non-unique stateIds. 
(3) Using the set of non-unique stateIds, introduce a new mapping of user-provided stateIds to store ID 

// stateId used in a single ParDo continue to use the same stateId as storeId
.apply("First stateful ParDo", ParDo.of(fn)); 
→ “foo"

// stateId used in multiple ParDos use a new mapping to store Id with PTransform name
.apply("Second stateful ParDo", ParDo.of(fn2)); 
→ “foo-Second_stateful_ParDo”

(4) Re-enable ParDoTest in :runners:samza build.gradle

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java [hunk 1]
```diff
@@ -35,12 +34,17 @@ public class ConfigContext {
   private final Map<PValue, String> idMap;
   private AppliedPTransform<?, ?, ?> currentTransform;
   private final SamzaPipelineOptions options;
-  private final Set<String> stateIds;
+  private final Map<String, String> usedStateIdMap;
+  private final Map<String, String> multiParDoStateIdMap;
```
Comment (dxichen) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java:38]: Lets rename this for clarity stateIdsToRewrite the value could also be the new state id.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java [hunk 2]
```diff
@@ -249,6 +252,19 @@ public String getTransformId() {
...
 
+  /** Given a set of user stateIds and parDo name, return a stateId to storeId map. */
+  public Map<String, String> getStateIdToStoreIdMap(Set<String> stateIds, String escapedParDoName) {
+    final Map<String, String> storeIds = new HashMap<>();
+    stateIds.forEach(
+        stateId ->
+            storeIds.put(
+                stateId,
+                multiParDoStateIds.contains(stateId)
+                    ? String.join("-", stateId, escapedParDoName)
```
Comment (dxichen) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java:263]: Let's create a helper method for this join, looks like it happens in a lot of place in the code.

**Review by dxichen — CHANGES_REQUESTED** [review]
Minor style/readability comments, thanks for the diff!

File: runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java [hunk 3]
```diff
@@ -140,9 +141,21 @@ public SamzaPipelineResult run(Pipeline pipeline) {
     LOG.info("Beam pipeline JSON graph:\n{}", jsonGraph);
 
     final Map<PValue, String> idMap = PViewToIdMapper.buildIdMap(pipeline);
+    /* Map of stateId to sanitized (remove whitespace and replace '-' with '_') PTransform name, used in multiple ParDos
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java]: Moved the javadoc with /**.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java [hunk 4]
```diff
@@ -105,6 +106,38 @@ public static void createConfig(
...
+          // replace old single parDo store configs with new storeId mapping appended with parDo
+          // name
+          configBuilder.remove("stores." + stateId + ".factory");
+          configBuilder.remove("stores." + stateId + ".key.serde");
+          configBuilder.remove("stores." + stateId + ".msg.serde");
+          configBuilder.remove("stores." + stateId + ".rocksdb.compression");
+          // put new config with multi pardo config
+          configBuilder.put(
+              "stores." + multiParDoStateId + ".factory",
+              "org.apache.samza.storage.kv.RocksDbKeyValueStorageEngineFactory");
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java:136]: Hi Daniel, could you elaborate more on what you mean by "already including the package?" Thanks! We do have some occurences of using these classnames in other places where we add the configs.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 5]
```diff
@@ -149,37 +139,39 @@ static <K> Factory<K> createStateInternalsFactory(
...
       SamzaPipelineOptions pipelineOptions,
       ExecutableStage executableStage) {
-
-    Set<String> stateIds =
+    // TODO: handle same stateIds in multiple ParDos for portable mode
+    Map<String, String> stateIds =
         executableStage.getUserStates().stream()
             .map(UserStateReference::localName)
-            .collect(Collectors.toSet());
+            .collect(Collectors.toMap(Function.identity(), Function.identity()));
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Thanks, made the change!

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java (hunk 1 above)
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java]: Sounds good - made the change.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java (hunk 2 above)
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java]: Done. Thanks!

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 6]
```diff
@@ -129,18 +130,7 @@ static KeyValueStore<ByteArray, StateValue<?>> getBeamStore(TaskContext context)
    */
   static <K> Factory<K> createNonKeyedStateInternalsFactory(
       String id, TaskContext context, SamzaPipelineOptions pipelineOptions) {
-    return createStateInternalsFactory(id, null, context, pipelineOptions, Collections.emptySet());
-  }
-
-  static <K> Factory<K> createStateInternalsFactory(
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:135]: As we fetch stateIds from DoFnSignature earlier in the logic for handling stateIds with multiParDo, I've replaced this with using stateIdMap (map of stateId to RocksDB storeId). Would we prefer to add this back for future use?

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dxichen) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java:136]: Oh I just mean something like `RocksDbKeyValueStorageEngineFactory.class.getName()`

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dxichen) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java:135]: got it, lets leave this out if it is no longer used

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java]: Ah Sounds good :) Made the change to all other occurrences.

**Review by xinyuiscool — COMMENTED** [review]
Two main issues:

1) no rewriting of configs, instead, do a prescan of pipeline and build the non-unique state id set
2) the changes in the SamzaStateInternals are not needed if you make the stores map still follow the same definition of {stateId -> store}. The state internals is generated per PTransform so it will not have dups here.

**Comment by seungjin-an** [comment]
> Two main issues:
> 
> 1. no rewriting of configs, instead, do a prescan of pipeline and build the non-unique state id set
> 2. the changes in the SamzaStateInternals are not needed if you make the stores map still follow the same definition of {stateId -> store}. The state internals is generated per PTransform so it will not have dups here.

Thank you so much for taking the time reviewing, providing suggestions on the pre-scan/ better naming, and fixing my understanding in SamzaStoreInternals. I have addressed all the comments and please have another look.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java [hunk 7]
```diff
@@ -35,12 +35,17 @@ public class ConfigContext {
   private final Map<PValue, String> idMap;
   private AppliedPTransform<?, ?, ?> currentTransform;
   private final SamzaPipelineOptions options;
-  private final Set<String> stateIds;
+  private final Map<String, String> usedStateIdMap;
+  private final Map<String, String> stateIdsToRewrite;
 
-  public ConfigContext(Map<PValue, String> idMap, SamzaPipelineOptions options) {
+  public ConfigContext(
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java:39]: Thank you for the suggestion here, made the change.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java [hunk 8]
```diff
@@ -140,9 +141,14 @@ public SamzaPipelineResult run(Pipeline pipeline) {
     LOG.info("Beam pipeline JSON graph:\n{}", jsonGraph);
 
     final Map<PValue, String> idMap = PViewToIdMapper.buildIdMap(pipeline);
+    final Map<String, String> multiParDoStateIdMap = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java]: I totally agree - this approach seems much better. Made the change, thank you!

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/DoFnOp.java [hunk 9]
```diff
@@ -148,7 +149,8 @@ public DoFnOp(
       JobInfo jobInfo,
       Map<String, TupleTag<?>> idToTupleTagMap,
       DoFnSchemaInformation doFnSchemaInformation,
-      Map<?, PCollectionView<?>> sideInputMapping) {
+      Map<?, PCollectionView<?>> sideInputMapping,
+      Map<String, String> userStateIds) {
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/DoFnOp.java]: As this is no longer needed, I reverted this diff but definitely agree. Thank you

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaDoFnRunners.java [hunk 10]
```diff
@@ -92,14 +90,14 @@ public static <InT, FnOutT> DoFnRunner<InT, FnOutT> create(
       DoFnSchemaInformation doFnSchemaInformation,
       Map<String, PCollectionView<?>> sideInputMapping,
       OpEmitter emitter,
-      FutureCollector futureCollector) {
+      FutureCollector futureCollector,
+      Map<String, String> userStateIds) {
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaDoFnRunners.java]: Sounds good

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 11]
```diff
@@ -99,13 +96,16 @@
 
   // the stores include both beamStore for system states as well as stores for user state
   private final Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores;
+  // Map of non-system stateIds to unique storeId
+  private final Map<String, String> stateIdMap;
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Sounds good.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 12]
```diff
@@ -149,37 +139,39 @@ static <K> Factory<K> createStateInternalsFactory(
...
     final Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores = new HashMap<>();
     stores.put(BEAM_STORE, getBeamStore(context));
-
     final Coder<K> stateKeyCoder;
     if (keyCoder != null) {
-      stateIds.forEach(
-          stateId ->
-              stores.put(
-                  stateId, (KeyValueStore<ByteArray, StateValue<?>>) context.getStore(stateId)));
+      stateIdMap
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Got it - thank you.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 13]
```diff
@@ -99,13 +96,16 @@
...
+  // Map of non-system stateIds to unique storeId
+  private final Map<String, String> stateIdMap;
   private final K key;
   private final byte[] keyBytes;
   private final int batchGetSize;
   private final String stageId;
 
   private SamzaStoreStateInternals(
       Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores,
+      Map<String, String> stateIds,
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Ack removed the changes in this class

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 14]
```diff
@@ -99,13 +96,16 @@
...
   private final Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores;
+  // Map of non-system stateIds to unique storeId
+  private final Map<String, String> stateIdMap;
   private final K key;
   private final byte[] keyBytes;
   private final int batchGetSize;
   private final String stageId;
 
   private SamzaStoreStateInternals(
       Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores,
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Ack removed the changes in this class

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 15]
```diff
@@ -149,37 +139,39 @@ static <K> Factory<K> createStateInternalsFactory(
...
     if (keyCoder != null) {
-      stateIds.forEach(
-          stateId ->
-              stores.put(
-                  stateId, (KeyValueStore<ByteArray, StateValue<?>>) context.getStore(stateId)));
+      stateIdMap
+          .values()
+          .forEach(
+              storeId ->
+                  stores.put(
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Ack removed the changes in this class

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 16]
```diff
@@ -149,37 +139,39 @@ static <K> Factory<K> createStateInternalsFactory(
...
+                      storeId,
+                      (KeyValueStore<ByteArray, StateValue<?>>) context.getStore(storeId)));
       stateKeyCoder = keyCoder;
     } else {
       stateKeyCoder = (Coder<K>) VoidCoder.of();
     }
-    return new Factory<>(Objects.toString(id), stores, stateKeyCoder, batchGetSize);
+    return new Factory<>(Objects.toString(id), stores, stateIdMap, stateKeyCoder, batchGetSize);
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 17]
```diff
@@ -271,16 +263,19 @@ private static ByteArrayOutputStream getThreadLocalBaos() {
   public static class Factory<K> implements StateInternalsFactory<K> {
     private final String stageId;
     private final Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores;
+    private final Map<String, String> stateIdMap;
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 18]
```diff
@@ -271,16 +263,19 @@ private static ByteArrayOutputStream getThreadLocalBaos() {
...
     private final Coder<K> keyCoder;
     private final int batchGetSize;
 
     public Factory(
         String stageId,
         Map<String, KeyValueStore<ByteArray, StateValue<?>>> stores,
+        Map<String, String> stateIdMap,
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 19]
```diff
@@ -271,16 +263,19 @@ private static ByteArrayOutputStream getThreadLocalBaos() {
...
         Coder<K> keyCoder,
         int batchGetSize) {
       this.stageId = stageId;
       this.stores = stores;
+      this.stateIdMap = stateIdMap;
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 20]
```diff
@@ -303,7 +298,8 @@ public StateInternals stateInternalsForKey(@Nullable K key) {
         throw new RuntimeException("Cannot encode key for state store", e);
       }
 
-      return new SamzaStoreStateInternals<>(stores, key, baos.toByteArray(), stageId, batchGetSize);
+      return new SamzaStoreStateInternals<>(
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 21]
```diff
@@ -323,11 +319,11 @@ protected AbstractSamzaState(
       this.coder = coder;
       this.namespace = namespace;
       this.addressId = address.getId();
-      this.isBeamStore = !stores.containsKey(address.getId());
+      this.isBeamStore = !stores.containsKey(stateIdMap.get(address.getId()));
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java [hunk 22]
```diff
@@ -323,11 +319,11 @@ protected AbstractSamzaState(
...
       this.store =
           isBeamStore
               ? (KeyValueStore) stores.get(BEAM_STORE)
-              : (KeyValueStore) stores.get(address.getId());
+              : (KeyValueStore) stores.get(stateIdMap.get(address.getId()));
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaStoreStateInternals.java]: Done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java [hunk 23]
```diff
@@ -83,6 +84,10 @@ public void putAll(Map<String, String> properties) {
     config.putAll(properties);
   }
 
+  public void remove(String name) {
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigBuilder.java]: Makes sense - done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java [hunk 24]
```diff
@@ -64,8 +69,39 @@ public SamzaPipelineOptions getPipelineOptions() {
...
+      return stateId;
+    } else {
+      // Same state id identified for the first time
+      if (!stateIdsToRewrite.containsKey(stateId)) {
+        final String prevParDoName = usedStateIdMap.get(stateId);
+        final String prevMultiParDoStateId =
+            StoreIdUtils.toMultiParDoStoreId(stateId, prevParDoName);
+        usedStateIdMap.put(prevMultiParDoStateId, prevParDoName);
+        // Store the stateId with previous parDo name which will be used for config rewriting
+        stateIdsToRewrite.put(stateId, prevParDoName);
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java]: Changed the logic to prescanning.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java [hunk 25]
```diff
@@ -377,25 +392,21 @@ public Map<String, String> createConfig(
...
-        // TODO: remove validation after we support same state id in different ParDo.
-        if (!ctx.addStateId(storeId)) {
-          throw new IllegalStateException(
-              "Duplicate StateId " + storeId + " found in multiple ParDo.");
-        }
-
+        final String userStateId = state.id();
+        final String escapedParDoName =
+            SamzaPipelineTranslatorUtils.escape(node.getEnclosingNode().getFullName());
+        final String uniqueStoreId = ctx.getUniqueStoreId(userStateId, escapedParDoName);
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java]: Only need config creation part for translation so removed this part.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java [hunk 26]
```diff
@@ -105,6 +108,46 @@ public static void createConfig(
...
+  /**
+   * Rewrite user store configs if there exists same state ids used in multiple ParDos. For each
+   * entry of a stateId to escaped PTransform name of first occurrence in topological traversal,
+   * rewrite RocksDB configs with the new mapping enforced from stateId to storeId.
+   * (eg) @StateId("foo") used in two ParDos fn, fn2: .apply("First Stateful ParDo with same
+   * stateId", ParDo.of(fn)) .apply("Second Stateful ParDo with same stateId", ParDo.of(fn2)) Map =
+   * ("foo", "First_Stateful_ParDo_with_same_stateId") storeId =
+   * "foo-First_Stateful_ParDo_with_same_stateId"
+   */
+  public static void rewriteConfigWithMultiParDoStateId(
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/SamzaPipelineTranslator.java]: Thank you :-)

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java [hunk 27]
```diff
@@ -83,6 +84,7 @@ public class TranslationContext {
   private final Map<PValue, String> idMap;
   private final Map<String, MessageStream> registeredInputStreams = new HashMap<>();
   private final Map<String, Table> registeredTables = new HashMap<>();
+  private final Set<String> multiParDoStateIds;
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java]: I really like "nonUniqueStateIds".. made the change throughout the diff. Thank you!

File: runners/samza/src/main/java/org/apache/beam/runners/samza/util/StoreIdUtils.java [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+ */
+package org.apache.beam.runners.samza.util;
+
+public class StoreIdUtils {
+
+  /**
+   * Join stateId and escaped PTransform name, used for RocksDB storeId of stateIds with multiple
+   * ParDos.
+   */
+  public static String toMultiParDoStoreId(String stateId, String escapedPTransformName) {
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/util/StoreIdUtils.java]: Done.

**Comment by xinyuiscool** [comment]
Thanks for fixing the problem for #1 in my previous comments. However, seems some of the necessary changes are also reverted for #2. I expect passing a mapping {stateId -> storeId} (currently it's a set of stateId) to the ```SamzaStoreStateInternals.createStateInternalsFactory()```. Then createStateInternalsFactory() can use that information to create the ```stores``` map.  Somehow this logic is missing. It was present in your previous code. Please take a look and run ./gradlew :runners:samza:build to verify the tests pass.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/DoFnOp.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/DoFnOp.java]: Made the change.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java]: Now that we removed the rewriting config logic, we no longer have "userStateIdToStoreIdMap". The change is that we have the nonUniqueStateId set from StateIdParser.scan(). Then, given this set, we'll create the stateIdToStoreMapping for each ParDo and initialize the SamzaStateStoreInternals, while keeping the stateId-> store map in "stores". Please take a look and let me know if this looks reasonable. Thank you!

**Comment by seungjin-an** [comment]
Thank you @xinyuiscool for the review again. SamzaStoreStateInternals logic has been included and verified the changes with all the tests passing from ./gradlew :runners:samza:build

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java [hunk 29]
```diff
@@ -35,12 +34,13 @@ public class ConfigContext {
   private final Map<PValue, String> idMap;
   private AppliedPTransform<?, ?, ?> currentTransform;
   private final SamzaPipelineOptions options;
-  private final Set<String> stateIds;
+  private final Set<String> nonUniqueStateIds;
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java:37]: Please take a look at the comments below to wrap this in a StoreIdGenerator class.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java [hunk 30]
```diff
@@ -83,6 +83,7 @@ public class TranslationContext {
   private final Map<PValue, String> idMap;
   private final Map<String, MessageStream> registeredInputStreams = new HashMap<>();
   private final Map<String, Table> registeredTables = new HashMap<>();
+  private final Set<String> nonUniqueStateIds;
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/TranslationContext.java:86]: See comments above to wrap this in a StoreIdGenerator class.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/util/StoreIdUtils.java [hunk 31]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.samza.util;
+
+public class StoreIdUtils {
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/util/StoreIdUtils.java:20]: We can move this method in this class into the new StoreIdGenerator.

**Review by xinyuiscool — COMMENTED** [review]
Overall looks great! Have some minor comments about consolidating some duplicate logic into a single class.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java [hunk 32]
```diff
@@ -377,17 +392,15 @@ public Map<String, String> createConfig(
...
       // set up user state configs
       for (DoFnSignature.StateDeclaration state : signature.stateDeclarations().values()) {
-        final String storeId = state.id();
-
-        // TODO: remove validation after we support same state id in different ParDo.
-        if (!ctx.addStateId(storeId)) {
-          throw new IllegalStateException(
-              "Duplicate StateId " + storeId + " found in multiple ParDo.");
+        final String stateId = state.id();
+        String storeId = stateId;
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java]: Awesome, changes are made.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java [hunk 33]
```diff
@@ -161,6 +163,17 @@ private static <InT, OutT> void doTranslate(
     Map<String, PCollectionView<?>> sideInputMapping =
         ParDoTranslation.getSideInputMapping(ctx.getCurrentTransform());
 
+    final DoFnSignature signature = DoFnSignatures.getSignature(transform.getFn().getClass());
+    final Map<String, String> stateIdToStoreMapping = new HashMap<>();
+    for (String stateId : signature.stateDeclarations().keySet()) {
+      String storeId = stateId;
```
[the remark this answers is not in the document]
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java]: Done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java (hunk 29 above)
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ConfigContext.java]: Done.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/util/StoreIdUtils.java (hunk 31 above)
Comment (seungjin-an) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/util/StoreIdUtils.java]: Done.

**Comment by seungjin-an** [comment]
> Overall looks great! Have some minor comments about consolidating some duplicate logic into a single class.

@xinyuiscool Awesome, I've addressed all the comments above. Please review upon your availability! Thank you.

**Review by xinyuiscool — APPROVED** [review]
LGTM!

**Comment by xinyuiscool** [comment]
Run Java PreCommit

**Pull request merged**